src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -175,11 +176,233 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
+                            indexerAccess is not null ? indexerAccess.Type :
+                            throw ExceptionUtilities.Unreachable;
+            }
+            else
+            {
+                if (!inputType.IsErrorType())
+                    diagnostics.Add(ErrorCode.ERR_UnsupportedTypeForSlicePattern, node.Location, inputType.ToDisplayString());
+                hasErrors = true;
+                sliceType = CreateErrorType();
+            }
+
+            BoundPattern? pattern = node.Pattern is not null
+                ? BindPattern(node.Pattern, sliceType, ExternalScope, permitDesignations, hasErrors, diagnostics)
+                : null;
+
+            return new BoundSlicePattern(node, sliceMethod, indexerAccess, sliceType: sliceType, pattern, inputType: inputType, narrowedType: inputType, hasErrors);
+        }
+
+        private ImmutableArray<BoundPattern> BindListPatternSubpatterns(
+            PropertyPatternClauseSyntax node,
+            TypeSymbol inputType,
+            TypeSymbol elementType,
+            bool permitDesignations,
+            ref bool hasErrors,
+            out bool sawSlice,
+            BindingDiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.IsKind(SyntaxKind.ListPatternClause));
+
+            sawSlice = false;
+            var subpatterns = node.Subpatterns;
+            var builder = ArrayBuilder<BoundPattern>.GetInstance(subpatterns.Count);
+            foreach (SubpatternSyntax subpat in subpatterns)
+            {
+                // Any name would have make this a PropertyPatternClause
+                Debug.Assert(subpat.NameColon is null);
+
+                PatternSyntax pattern = subpat.Pattern;
+                BoundPattern boundPattern;
+                if (pattern is SlicePatternSyntax slice)
+                {
+                    if (sawSlice)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_MisplacedSlicePattern, slice.Location);
+                        hasErrors = true;
+                    }
+
+                    boundPattern = BindSlicePattern(slice, inputType, permitDesignations, hasErrors, misplaced: false, diagnostics);
+                    sawSlice = true;
+                }
+                else
+                {
+                    boundPattern = BindPattern(pattern, elementType, ExternalScope, permitDesignations, hasErrors, diagnostics);
+                }
+
+                builder.Add(boundPattern);
+            }
+
+            return builder.ToImmutableAndFree();
+        }
+
+        private BoundPattern? BindLengthPatternClause(RecursivePatternSyntax node, TypeSymbol inputType,
+            bool permitDesignations, out PropertySymbol? lengthProperty, ref bool hasErrors, BindingDiagnosticBag diagnostics)
+        {
+            lengthProperty = null;
+
+            if (node.LengthPatternClause is not null || node.PropertyPatternClause.IsKind(SyntaxKind.ListPatternClause))
```

Comment:
>node.PropertyPatternClause.IsKind(SyntaxKind.ListPatternClause) [](http://example.com/codeflow?start=56&length=63)

I don't really understand why we're doing this if there's a list pattern and no length pattern.

Reply:
List pattern itself requires the length. I'd agree we should move this out of BindLengthPatternClause for clarity.